File: frontend-web/src/app/core/services/promociones.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { environment } from '../../../environments/environment';
import { Promocion, PromocionCreate, PromocionUpdate } from '../models/promocion.model';

@Injectable({ providedIn: 'root' })
export class PromocionesService {
  private http = inject(HttpClient);
  private readonly base = `${environment.apiUrl}/promociones`;

  listar(soloActivas = false) {
    return this.http.get<Promocion[]>(this.base, { params: { solo_activas: soloActivas } });
  }

  obtener(id: number) {
    return this.http.get<Promocion>(`${this.base}/${id}`);
  }

  crear(data: PromocionCreate) {
    return this.http.post<Promocion>(this.base, data);
  }

  actualizar(id: number, data: PromocionUpdate) {
    return this.http.put<Promocion>(`${this.base}/${id}`, data);
  }

  desactivar(id: number) {
    return this.http.delete<Promocion>(`${this.base}/${id}`);
  }

  reactivar(id: number) {
    return this.http.patch<Promocion>(`${this.base}/${id}/reactivar`, {});
  }

  actualizarProductos(id: number, productoIds: number[]) {
    return this.http.put<Promocion>(`${this.base}/${id}/productos`, { producto_ids: productoIds });
  }
}

File: mobile-app/lib/features/admin/promociones/services/promociones_service.dart
import '../../../../core/network/dio_client.dart';
import '../models/promocion.dart';

class PromocionesService {
  final DioClient _client = DioClient();

  /// Obtiene la lista de promociones (filtrable por activas).
  Future<List<Promocion>> listar({bool soloActivas = false}) async {
    final response = await _client.get(
      '/promociones',
      query: {'solo_activas': soloActivas},
    );
    final list = response.data as List<dynamic>? ?? [];
    return list
        .map((e) => Promocion.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  /// Obtiene una promoción por ID con su detalle.
  Future<Promocion> obtener(int id) async {
    final response = await _client.get('/promociones/$id');
    return Promocion.fromJson(response.data as Map<String, dynamic>);
  }

  /// Crea una nueva promoción.
  Future<Promocion> crear(Map<String, dynamic> datos) async {
    final response = await _client.post('/promociones', data: datos);
    return Promocion.fromJson(response.data as Map<String, dynamic>);
  }

  /// Actualiza una promoción existente.
  Future<Promocion> actualizar(int id, Map<String, dynamic> datos) async {
    final response = await _client.put('/promociones/$id', data: datos);
    return Promocion.fromJson(response.data as Map<String, dynamic>);
  }

  /// Desactiva una promoción (soft-delete).
  Future<Promocion> desactivar(int id) async {
    final response = await _client.delete('/promociones/$id');
    return Promocion.fromJson(response.data as Map<String, dynamic>);
  }

  /// Reactiva una promoción desactivada.
  Future<Promocion> reactivar(int id) async {
    final response = await _client.patch('/promociones/$id/reactivar');
    return Promocion.fromJson(response.data as Map<String, dynamic>);
  }

  /// Actualiza la lista de productos asociados.
  Future<Promocion> actualizarProductos(int id, List<int> productoIds) async {
    final response = await _client.put(
      '/promociones/$id/productos',
      data: {'producto_ids': productoIds},
    );
    return Promocion.fromJson(response.data as Map<String, dynamic>);
  }
}
